src/pybind/mgr/object_format.py

```diff
@@ -0,0 +1,272 @@
+# object_format.py provides types and functions for working with
+# requested output formats such as JSON, YAML, etc.
+
+import enum
+import errno
+import json
+
+from functools import wraps
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+try:
+    from typing import Protocol  # Protocol was added in Python 3.8
+except ImportError:
+
+    class Protocol:  # type: ignore
+        pass
+
+
+import yaml
+
+from mgr_module import HandlerFuncType
+
+
+class Format(enum.Enum):
+    plain = "plain"
+    json = "json"
+    json_pretty = "json-pretty"
+    yaml = "yaml"
+    xml_pretty = "xml-pretty"
+    xml = "xml"
+
+
+def _obj_to_json(obj: Any) -> Any:
+    tj = getattr(obj, "to_json", None)
+    if tj:
+        return tj()
+    return obj
+
+
+def _obj_to_yaml(obj: Any) -> Any:
+    if hasattr(obj, "yaml_representer"):
+        return obj
+    return _obj_to_json(obj)
+
+
+def to_json(
+    obj: Any, *, many: bool = False, indent: Optional[int] = None
+) -> str:
+    """Return a JSON string representation of the given object `obj`.
+    If `many` is true, treat `obj` as a collection of objects to be
+    serialized.
+    """
+    if many:
+        json_obj = [_obj_to_json(o) for o in obj]
+    else:
+        json_obj = _obj_to_json(obj)
+    return json.dumps(json_obj, indent=indent, sort_keys=True)
+
+
+def to_yaml(
```

Comment:
I'd suggest putting both `to_json` and `_obj_to_json` under a common class or scope (or a `SimpleNamespace`).

If you want ideas for the naming: in the C++ side, the output format is handled by `Formatter` specialized classes: `JSONFormatter`, `PyFormatter`, `XMLFormatter`, `HTMLFormatter` (see `src/common/formatter.h`).

Reply:
What's the objective of putting these functions into a class? I don't see the value-add. When I write python I always avoid writing a class if a simple function will do.

Reply:
Why putting this code in a new file and not in an existing (unrelated) one?

Because they're related (`_obj_to_json` and `to_json`), and there's some implicit subordination between them (the leading `_` might suggest some kind of 'protected' access to that, so not meant to be externally consumed).

By using classes you might also feel tempted of making them reusable (can the Yaml formatter reuse code from the JSON one or a common parent class? There seems to be some code duplication over there).

Just by looking at something called `YAMLFormatter` I may understand what's this file about, but with `_obj_to_json` it's harder for me to get what's the final purpose of this.

And finally, if someone has to unit-test and mock this, it's easier to mock a set of related methods by mocking a whole class, rather than cherry-picking methods from a file.